contracts/GnosisSafe.sol

```diff
@@ -235,14 +237,9 @@ contract GnosisSafe
                 require(ISignatureValidator(currentOwner).isValidSignature(data, contractSignature) == EIP1271_MAGIC_VALUE, "Invalid contract signature provided");
             // If v is 1 then it is an approved hash
             } else if (v == 1) {
-                // When handling approved hashes the address of the approver is encoded into r
-                currentOwner = address(uint256(r));
-                // Hashes are automatically approved by the sender of the message or when they have been pre-approved via a separate transaction
-                require(msg.sender == currentOwner || approvedHashes[currentOwner][dataHash] != 0, "Hash has not been approved");
-                // Hash has been marked for consumption. If this hash was pre-approved free storage
-                if (consumeHash && msg.sender != currentOwner) {
-                    approvedHashes[currentOwner][dataHash] = 0;
-                }
+                // When handling approved hashes the address position on approvedHashes array of the approver is encoded into r
+                currentOwner = approvedHashes[dataHash][uint256(r)];
```

Comment:
This changes the requirements for using the pre-approved transactions to require submitting additional "helper" data along with the "pre-approved" signature type. Consider this for its effects downstream.